src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -238,6 +238,69 @@ impl ScalarFunc {
         let week = t.get_time().iso_week().week();
         Ok(Some(i64::from(week)))
     }
+
+    #[inline]
+    pub fn add_datetime_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let d: Cow<'a, MyDuration> = try_opt_or!(
+            self.children[1].eval_duration(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let add = t
+            .get_time()
+            .checked_add_signed(Duration::nanoseconds(d.to_nanos()));
+        if add.is_none() {
+            return Err(box_err!("parse from duration {} overflows", d));
+        }
+        let mut res = t.to_mut().clone();
+        res.set_time(add.unwrap());
```

Comment:
This `unwrap` call can be avoided by changing the previous `if add.is_none()` to a pattern matching.

Reply:
@ice1000 Done.